File: Mesh4n/trunk/Main/Source/SyncEngine.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Mesh4n
{
	/// <summary>
	/// Main class that performs synchronization between two repositories.
	/// </summary>
	public class SyncEngine
	{
		public event EventHandler<ItemEventArgs> ItemReceived;
		public event EventHandler<ItemEventArgs> ItemSent;

		IRepositoryAdapter source;
		IRepositoryAdapter target;

		/// <summary>
		/// Initializes the engine with the two repositories to synchronize.
		/// </summary>
		public SyncEngine(IRepositoryAdapter source, IRepositoryAdapter target)
		{
			Guard.ArgumentNotNull(source, "left");
			Guard.ArgumentNotNull(target, "right");

			this.source = source;
			this.target = target;
		}

		/// <summary>
		/// Performs a full sync between the two repositories, automatically 
		/// incorporating changes in both.
		/// </summary>
		/// <remarks>
		/// Items on the source repository are sent first, and then the 
		/// changes from the target repository are incorporated into the source.
		/// </remarks>
		/// <returns>The list of items that had conflicts.</returns>
		public IList<Item> Synchronize()
		{
			return SynchronizeImpl(null, new MergeFilter(), new ItemFilter());
		}

		/// <summary>
		/// Performs a full sync between the two repositories, optionally calling the 
		/// given <paramref name="mergeFilter"/> callback.
		/// </summary>
		/// <remarks>
		/// Items on the source repository are sent first, and then the 
		/// changes from the target repository are incorporated into the source.
		/// </remarks>
		/// <returns>The list of items that had conflicts.</returns>
		public IList<Item> Synchronize(MergeFilter mergeFilter)
		{
			return SynchronizeImpl(null, mergeFilter, new ItemFilter());
		}

		/// <summary>
		/// Performs a full sync between the two repositories, automatically 
		/// incorporating changes in both.
		/// </summary>
		/// <param name="itemFilter">Represents a method that filter items according to a specified criteria. </param>
		/// <remarks>
		/// Items on the source repository are sent first, and then the 
		/// changes from the target repository are incorporated into the source.
		/// </remarks>
		/// <returns>The list of items that had conflicts.</returns>
		public IList<Item> Synchronize(ItemFilter itemFilter)
		{
			return SynchronizeImpl(null, new MergeFilter(), itemFilter);
		}

		/// <summary>
		/// Performs a partial sync between the two repositories since the specified date, automatically 
		/// incorporating changes in both.
		/// </summary>
		/// <param name="since">Synchronize changes that happened after this date.</param>
		/// <remarks>
		/// Items on the source repository are sent first, and then the 
		/// changes from the target repository are incorporated into the source.
		/// </remarks>
		/// <returns>The list of items that had conflicts.</returns>
		public IList<Item> Synchronize(DateTime? since)
		{
			return SynchronizeImpl(since, new MergeFilter(), new ItemFilter());
		}

		/// <summary>
		/// Performs a partial sync between the two repositories since the specified date, automatically 
		/// incorporating changes in both.
		/// </summary>
		/// <param name="since">Synchronize changes that happened after this date.</param>
		/// <param name="itemFilter">Represents a method that filter items according to a specified criteria.</param>
		/// <remarks>
		/// Items on the source repository are sent first, and then the 
		/// changes from the target repository are incorporated into the source.
		/// </remarks>
		/// <returns>The list of items that had conflicts.</returns>
		public IList<Item> Synchronize(DateTime? since, ItemFilter itemFilter)
		{
			return SynchronizeImpl(since, new MergeFilter(), itemFilter);
		}

		/// <summary>
		/// Performs a partial sync between the two repositories since the specified date, optionally calling the 
		/// given <paramref name="mergeFilter"/> callback.
		/// </summary>
		/// <param name="since">Synchronize changes that happened after this date.</param>
		/// <remarks>
		/// Items on the source repository are sent first, and then the 
		/// changes from the target repository are incorporated into the source.
		/// </remarks>
		/// <returns>The list of items that had conflicts.</returns>
		public IList<Item> Synchronize(DateTime? since, MergeFilter mergeFilter)
		{
			return SynchronizeImpl(since, mergeFilter, new ItemFilter());
		}

		/// <summary>
		/// Performs a partial sync between the two repositories since the specified date, optionally calling the 
		/// given <paramref name="mergeFilter"/> callback.
		/// </summary>
		/// <param name="since">Synchronize changes that happened after this date.</param>
		/// <param name="itemFilter">Represents a method that filter items according to a specified criteria.</param>
		/// <remarks>
		/// Items on the source repository are sent first, and then the 
		/// changes from the target repository are incorporated into the source.
		/// </remarks>
		/// <returns>The list of items that had conflicts.</returns>
		public IList<Item> Synchronize(DateTime? since, MergeFilter mergeFilter, ItemFilter itemFilter)
		{
			return SynchronizeImpl(since, mergeFilter, itemFilter);
		}

		private IList<Item> SynchronizeImpl(DateTime? since, MergeFilter mergeFilter, ItemFilter itemFilter)
		{
			Guard.ArgumentNotNull(mergeFilter, "mergeFilter");
			Guard.ArgumentNotNull(itemFilter, "itemFilter");

			IEnumerable<Item> outgoingItems = EnumerateItemsProgress(
				(since == null) ? source.GetAll(itemFilter.Left) : source.GetAllSince(since, itemFilter.Left),
				RaiseItemSent);

			if (!target.SupportsMerge)
			{
				IEnumerable<ItemMergeResult> outgoingToMerge = MergeItems(outgoingItems, target);
				if ((mergeFilter.Behaviors & MergeFilterBehaviors.Right) == MergeFilterBehaviors.Right)
				{
					outgoingToMerge = mergeFilter.Handler(target, outgoingToMerge);
				}
				Import(outgoingToMerge, target);
			}
			else
			{
				target.Merge(outgoingItems);
			}

			IEnumerable<Item> incomingItems = EnumerateItemsProgress(
				(since == null) ? target.GetAll(itemFilter.Right) : target.GetAllSince(since, itemFilter.Right),
				RaiseItemReceived);

			if (!source.SupportsMerge)
			{
				IEnumerable<ItemMergeResult> incomingToMerge = MergeItems(incomingItems, source);
				if ((mergeFilter.Behaviors & MergeFilterBehaviors.Left) == MergeFilterBehaviors.Left)
				{
					incomingToMerge = mergeFilter.Handler(source, incomingToMerge);
				}
				
				return Import(incomingToMerge, source);
			}
			else
			{
				// If repository supports its own SSE merge behavior, don't apply it locally.
				return new List<Item>(source.Merge(incomingItems));
			}
		}

		private IEnumerable<ItemMergeResult> MergeItems(IEnumerable<Item> items, IRepositoryAdapter repository)
		{
			foreach (Item incoming in items)
			{
				Item original = repository.Get(incoming.Sync.Id);
				ItemMergeResult result = new MergeBehavior().Merge(original, incoming);

				if (result.Operation != MergeOperation.None)
					yield return result;
			}
		}

		private IList<Item> Import(IEnumerable<ItemMergeResult> items, IRepositoryAdapter repository)
		{
			// Straight import of data in merged results. 
			// Conflicting items are saved and also 
			// are returned for conflict resolution by the user or 
			// a custom component. MergeBehavior determines 
			// the winner element that is saved.
			// Conflicts are returned in a list because we need 
			// the full iteration over the merged items to be 
			// processed. If we returned an IEnumerable, we would 
			// depend on the client iterating it in order to 
			// actually import items, which is undesirable.
			List<Item> conflicts = new List<Item>();

			foreach (ItemMergeResult result in items)
			{
				if (result.Operation != MergeOperation.None &&
					result.Proposed != null &&
					result.Proposed.Sync.Conflicts.Count > 0)
				{
					conflicts.Add(result.Proposed);
				}

				switch (result.Operation)
				{
					case MergeOperation.Added:
						// Clean history before adding
						repository.Add(
							new Item(result.Proposed.XmlItem, 
							Behaviors.SparsePurge(result.Proposed.Sync)));
						break;
					case MergeOperation.Updated:
					case MergeOperation.Conflict:
						// Clean history before updating
						repository.Update(
							new Item(result.Proposed.XmlItem,
							Behaviors.SparsePurge(result.Proposed.Sync)));
						break;
					case MergeOperation.None:
						break;
					default:
						throw new InvalidOperationException();
				}
			}

			return conflicts;
		}

		private IEnumerable<Item> EnumerateItemsProgress(IEnumerable<Item> items, RaiseHandler raiser)
		{
			foreach (Item item in items)
			{
				raiser(item);
				yield return item;
			}
		}

		private void RaiseItemReceived(Item item)
		{
			if (ItemReceived != null)
				ItemReceived(this, new ItemEventArgs(item));
		}

		private void RaiseItemSent(Item item)
		{
			if (ItemSent != null)
				ItemSent(this, new ItemEventArgs(item));
		}

		delegate void RaiseHandler(Item item);
	}
}

File: Mesh4j2me/tags/Mesh4x_2_0_0/org.mesh4j.sync/src/org/mesh4j/sync/SyncEngine.java
package org.mesh4j.sync;

import java.util.Date;
import java.util.Vector;

import org.mesh4j.sync.merge.MergeBehavior;
import org.mesh4j.sync.merge.MergeResult;
import org.mesh4j.sync.model.Item;
import org.mesh4j.sync.observer.IObserverItem;
import org.mesh4j.sync.observer.ObservableItem;
import org.mesh4j.sync.validations.Guard;


/**
 * Main class that performs synchronization between two repositories.
 * 
 * @author jtondato
 */
public class SyncEngine {

	// MODEL VARIABLES
	private ObservableItem itemReceived = new ObservableItem();
	private ObservableItem itemSent = new ObservableItem();

	private ISyncAdapter source;
	private ISyncAdapter target;

	// BUSINESS METHODS
	public SyncEngine(ISyncAdapter source, ISyncAdapter target) {
		super();

		Guard.argumentNotNull(source, "left");
		Guard.argumentNotNull(target, "right");

		this.source = source;
		this.target = target;
	}

	// / <summary>
	// / Performs a full sync between the two repositories, automatically
	// / incorporating changes in both.
	// / </summary>
	// / <remarks>
	// / Items on the source repository are sent first, and then the
	// / changes from the target repository are incorporated into the source.
	// / </remarks>
	// / <returns>The list of items that had conflicts.</returns>
	public Vector<Item> synchronize() {
		return synchronize(null, NullPreviewHandler.INSTANCE,
				PreviewBehavior.None);
	}

	// / <summary>
	// / Performs a full sync between the two repositories, optionally calling
	// the
	// / given <paramref name="previewer"/> callback as specified by the
	// <paramref name="behavior"/> argument.
	// / </summary>
	// / <remarks>
	// / Items on the source repository are sent first, and then the
	// / changes from the target repository are incorporated into the source.
	// / </remarks>
	// / <returns>The Vector of items that had conflicts.</returns>
	public Vector<Item> synchronize(IPreviewImportHandler previewer,
			PreviewBehavior behavior) {
		return synchronize(null, previewer, behavior);
	}

	// / <summary>
	// / Performs a partial sync between the two repositories since the
	// specified date, automatically
	// / incorporating changes in both.
	// / </summary>
	// / <param name="since">Synchronize changes that happened after this
	// date.</param>
	// / <remarks>
	// / Items on the source repository are sent first, and then the
	// / changes from the target repository are incorporated into the source.
	// / </remarks>
	// / <returns>The Vector of items that had conflicts.</returns>
	public Vector<Item> synchronize(Date since) {
		return synchronize(since, NullPreviewHandler.INSTANCE,
				PreviewBehavior.None);
	}

	// / <summary>
	// / Performs a partial sync between the two repositories since the
	// specified date, optionally calling the
	// / given <paramref name="previewer"/> callback as specified by the
	// <paramref name="behavior"/> argument.
	// / </summary>
	// / <param name="since">Synchronize changes that happened after this
	// date.</param>
	// / <remarks>
	// / Items on the source repository are sent first, and then the
	// / changes from the target repository are incorporated into the source.
	// / </remarks>
	// / <returns>The Vector of items that had conflicts.</returns>
	public Vector<Item> synchronize(Date since, IPreviewImportHandler previewer,
			PreviewBehavior behavior) {

		Guard.argumentNotNull(previewer, "previewer");

		this.beginSync();
		Vector<Item> result = null;
		Vector<Item> sourceItems = (since == null) ? source.getAll() : source.getAllSince(since);
		Vector<Item> outgoingItems = this.enumerateItemsProgress(sourceItems, this.itemSent);

		if (target instanceof ISupportMerge) {
			ISupportMerge targetMerge = (ISupportMerge) target;
			targetMerge.merge(outgoingItems);
		} else {
			Vector<MergeResult> outgoingToMerge = this.mergeItems(outgoingItems, target);
			if (behavior == PreviewBehavior.Right || behavior == PreviewBehavior.Both) {
				outgoingToMerge = previewer.preview(target, outgoingToMerge);
			}
			this.importItems(outgoingToMerge, target);
		}

		Vector<Item> targetItmes = (since == null) ? target.getAll() : target.getAllSince(since);
		Vector<Item> incomingItems = this.enumerateItemsProgress(targetItmes, this.itemReceived);

		if (source instanceof ISupportMerge) {
			// If repository supports its own SSE merge behavior, don't apply it locally.
			ISupportMerge sourceMerge = (ISupportMerge) source;
			result = sourceMerge.merge(incomingItems);				
		} else {
			Vector<MergeResult> incomingToMerge = this.mergeItems(incomingItems, source);
			if (behavior == PreviewBehavior.Left || behavior == PreviewBehavior.Both) {
				incomingToMerge = previewer.preview(source, incomingToMerge);
			}

			result = this.importItems(incomingToMerge, source);
		}
		this.endSync();	
			
		return result;
	}

	private Vector<MergeResult> mergeItems(Vector<Item> items,
			ISyncAdapter repository) {

		Vector<MergeResult> mergeResult = new Vector<MergeResult>();
		for (Item incoming : items) {
			Item original = repository.get(incoming.getSyncId());
			MergeResult result = MergeBehavior.merge(original, incoming);

			if (!result.isMergeNone()) {
				mergeResult.addElement(result);
			}
		}
		return mergeResult;
	}

	private Vector<Item> importItems(Vector<MergeResult> items,
			ISyncAdapter repository) {
		// Straight import of data in merged results.
		// Conflicting items are saved and also
		// are returned for conflict resolution by the user or
		// a custom component. MergeBehavior determines
		// the winner element that is saved.
		// Conflicts are returned in a Vector because we need
		// the full iteration over the merged items to be
		// processed. If we returned an IEnumerable, we would
		// depend on the client iterating it in order to
		// actually import items, which is undesirable.
		Vector<Item> conflicts = new Vector<Item>();

		for (MergeResult result : items) {			
			if (!result.isMergeNone() && result.getProposed() != null
					&& result.getProposed().hasSyncConflicts()) {
				conflicts.addElement(result.getProposed());
			}

			if (result.getOperation() == null
					|| result.getOperation().isRemoved()) {
				Guard.throwsException("UnsupportedOperation");						// TODO (JMT) resource bundle UnsupportedOperation
			} else if (result.getOperation().isAdded()) {
				repository.add(result.getProposed());
			} else if (result.getOperation().isUpdated()
					|| result.getOperation().isConflict()) {
				repository.update(result.getProposed());
			}
		}

		return conflicts;
	}

	private Vector<Item> enumerateItemsProgress(Vector<Item> items,
			ObservableItem observable) {
		Vector<Item> result = new Vector<Item>();
		for (Item item : items) {
			result.addElement(item);
			observable.notifyObservers(item);
		}
		return result;
	}

	public void registerItemReceivedObserver(IObserverItem ... observers) {
		for (IObserverItem itemObserver : observers) {
			this.itemReceived.addObserver(itemObserver);	
		}
	}

	public void removeItemReceivedObserver(IObserverItem ... observers) {
		for (IObserverItem itemObserver : observers) {
			this.itemReceived.removeObserver(itemObserver);
		}
	}

	public void registerItemSentObserver(IObserverItem ... observers) {
		for (IObserverItem itemObserver : observers) {
			this.itemSent.addObserver(itemObserver);
		}
	}

	public void removeItemSentObserver(IObserverItem ... observers) {
		for (IObserverItem itemObserver : observers) {
			this.itemSent.removeObserver(itemObserver);
		}
	}
	
	private void beginSync(){
		if(this.source instanceof ISyncAware){
			((ISyncAware)this.source).beginSync();
		}
		
		if(this.target instanceof ISyncAware){
			((ISyncAware)this.target).beginSync();
		}
	}

	private void endSync(){
		if(this.source instanceof ISyncAware){
			((ISyncAware)this.source).endSync();
		}
		
		if(this.target instanceof ISyncAware){
			((ISyncAware)this.target).endSync();
		}		
	}
}
